### ISSUE TITLE
Add test for Profiler timings

### ISSUE THREAD

**Issue by ananthsub** [PyTorchLightning/pytorch-lightning#4672]
## 🐛 Bug

Does the lightning simpler profiler do sampling? my training run took ~2 hours but the total times don't match up at all with the elapsed wall clock time: https://gist.github.com/ananthsub/67d93a7fefb359813447d94d8d64ebd4

## Please reproduce using [the BoringModel and post here](https://colab.research.google.com/drive/1p-REpS5ojxzz_GtfieOSyzZ_FADCE4Hm?usp=sharing)

https://colab.research.google.com/drive/1p-REpS5ojxzz_GtfieOSyzZ_FADCE4Hm?usp=sharing

### Expected behavior

Profiler results would be more representative of the actual training time

**Comment by Borda** [comment]
@ananthsub just to tried your sample code and observed the following differences for wrapped fit in 
```python
# Train the model ⚡
    t0 = time.monotonic()
    trainer.fit(model, train, val)
    dt = time.monotonic() - t0
```
and sum all profiler events:
```python
print (f'fit time: {time_fit}')
time_prof = sum([sum(t) for a, t in trainer.profiler.recorded_durations.items()])
print (f'profiler time: {time_prof}')
```
yield in 
```
fit time: 40.626095813000006
profiler time: 33.77439123699514
```
how was your discrepancies, even larger?

**Comment by Borda** [comment]
@jeremyjordan mind comment on it? I heard you about our testing with sleep which has 20% imprecission depending on OS, but this case is that all particular times shall sum to the same overall time, which seems to have now about +17% -10% difference...

**Comment by jeremyjordan** [comment]
hi @ananthsub , the profiler works by wrapping certain code in a context manager which captures the start and end times.

pseudo code
```
try:                                               # <-- enter the context manager
    start = time.monotonic()
    yield action_name                  # <-- code executes
finally:
    end = time.monotonic()        # <-- exit the context manager
```

which ends up being used at various parts of the training loop

```
with profiler.profile('on_batch_start'):
    # execute on_batch_start callbacks and model hooks
```

you can see the code [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/profiler/profilers.py). i don't know of a good way to test for accuracy of this simple profiler. right now, we're using `time.sleep()` but this doesn't give us an exact duration to measure against, so we end up with a relative tolerance of 20% in the test (see test [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/tests/test_profiler.py#L51)).

if there is an error, it will accumulate over each epoch. however, i wouldn't expect that you would need to profile for more than 1 epoch to get an understanding of where your code is being bottlenecked.

if you want more precise measurements, i recommend checking out the [AdvancedProfiler](https://pytorch-lightning.readthedocs.io/en/stable/profiler.html#advanced-profiling) object which uses Python's built-in cProfiler for measurement.

**Comment by jeremyjordan** [comment]
please also note that the total training time won't exactly line up with the sum of profiled events. we don't profile every line of code in the train loop, but rather focus on identifying bottlenecks in user-provided code (where you can actually make improvements to speed things up)

**Comment by Borda** [comment]
@ananthsub mind share how much time was missing in your 2-hour run?
